transaction.go

```diff
@@ -0,0 +1,489 @@
+	data := make([]byte, 32+len(amount))
+	i := 0
+	i += copy(data[i:], amount)
+	i += copy(data[i:], io.Address[:])
+	return data, nil
+}
+
+// MarshalBinary marshals a transaction signature to its binary representation. See
+// UnmarshalBinary for encoding details.
+func (s *FactoidTransactionSignature) MarshalBinary() ([]byte, error) {
+	if !s.IsPopulated() {
+		return nil, fmt.Errorf("not populated")
+	}
+
+	rcdData, err := s.ReedeemCondition.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	data := make([]byte, len(rcdData)+len(s.SignatureBlock))
+	i := 0
+	i += copy(data[i:], rcdData)
+	i += copy(data[i:], s.SignatureBlock)
+	return data, nil
+}
+
+const (
+	TransactionHeadMinLen = 0 + // Version length is varint
+		6 + // timestamp
+		1 + // input count
+		1 + // factoid output count
+		1 // EC output count
+
+	TransactionTotalMinLen = TransactionHeadMinLen // Coinbases have no body
+
+)
+
+// Decode will consume as many bytes as necessary to unmarshal the factoid
+// transaction. It will return the number of bytes read and an error.
+func (f *FactoidTransaction) Decode(data []byte) (i int, err error) {
+	// Because the length of an FactoidTransaction is hard to define up front, we will catch
+	// any sort of out of bound errors in a recover
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("failed to unmarshal")
+		}
+	}()
+
+	if len(data) < TransactionTotalMinLen {
+		return 0, fmt.Errorf("insufficient length")
+	}
+
+	// Decode header
+	i, err = f.DecodeHeader(data)
+	if err != nil {
+		return 0, err
+	}
+
+	// Decode the inputs
+	f.FCTInputs = make([]FactoidTransactionIO, f.InputCount)
+	read, err := FactoidTransactionIOs(f.FCTInputs).Decode(data[i:])
+	if err != nil {
+		return 0, err
+	}
+	i += read
+
+	// Decode the FCT Outputs
+	f.FCTOutputs = make([]FactoidTransactionIO, f.FCTOutputCount)
+	read, err = FactoidTransactionIOs(f.FCTOutputs).Decode(data[i:])
+	if err != nil {
+		return 0, err
+	}
+	i += read
+
+	// Decode the EC Outputs
+	f.ECOutputs = make([]FactoidTransactionIO, f.ECOutputCount)
+	read, err = FactoidTransactionIOs(f.ECOutputs).Decode(data[i:])
+	if err != nil {
+		return 0, err
+	}
+	i += read
+
+	// Decode the signature blocks, one per input
+	f.Signatures = make([]FactoidTransactionSignature, f.InputCount)
+	for c := uint8(0); c < f.InputCount; c++ {
+		// f.Signatures[i] = new(FactoidTransactionSignature)
+		read, err := f.Signatures[c].Decode(data[i:])
+		if err != nil {
+			return 0, err
+		}
+		i += read
+	}
+
+	txid, err := f.ComputeTransactionID()
+	if err != nil {
+		return 0, err
+	}
+	f.TransactionID = &txid
+
+	return i, err
+}
+
+// UnmarshalBinary unmarshals the data into a factoid transaction.
+func (f *FactoidTransaction) UnmarshalBinary(data []byte) error {
+	// TODO: Some length checks to prevent too few/too many bytes
+	_, err := f.Decode(data)
+	return err
+}
+
+func (f *FactoidTransaction) DecodeHeader(data []byte) (int, error) {
```

Comment:
Lets move this directly into Decode. This does not abstract something that gets reused elsewhere, so it forces the reader to visit another section of code for not a lot of good reason. Do this inline in Decode.